project1/server/file_handler/tests/test_thread_safe_file_handler.cpp

```diff
@@ -0,0 +1,222 @@
+#include <cstdint>
+#include <ctime>
+#include <filesystem>
+#include <memory>
+#include <random>
+#include <sstream>
+#include <thread>
+#include <vector>
+
+#include "../file_access_manager.h"
+#include "../thread_safe_file_handler.h"
+#include "gtest/gtest.h"
+
+using std::filesystem::path;
+
+namespace server::file_handler::tests {
+namespace {
+
+/**
+ * @brief Manages a temp directory to use for testing.
+ */
+class TestDir {
+ public:
+  TestDir() {
+    // Create the test directory.
+    const path kTempDir = path("/tmp");
+
+    std::default_random_engine generator;
+    std::uniform_int_distribution<uint32_t> distribution(0, 1000);
+
+    std::stringstream dir_name;
+    dir_name << "test_" << time(nullptr) << "_" << distribution(generator);
+    test_dir_ = kTempDir / dir_name.str();
+    std::filesystem::create_directory(test_dir_);
+  }
+
+  ~TestDir() {
+    // Remove the test directory.
+    std::filesystem::remove_all(test_dir_);
+  }
+
+  /**
+   * @return The path to the testing directory.
+   */
+  [[nodiscard]] path Get() const { return test_dir_; }
+
+ private:
+  /// Directory to use for storing the tests.
+  path test_dir_;
+};
+
+/**
+ * @brief Encapsulates the FileAccessManagers we are using for these tests.
+ */
```

Comment:
So, the idea here is that all threads that would perform a `file_handler` operation will share the same `AccessManagers` ?

Reply:
Yes, that is a necessity.